Prevent new dead Haskell code

sellout opened the pull request:
## Overview

This (re?)enables Weeder, which does whole-program dead-code analysis. No Haskell code is touched by this PR.

There are at least three different areas of the code base where I’ve discovered myself fixing code that isn’t referenced anywhere. While I fixed those areas upon encountering them, having dead code around adds noise and introduces confusion like I’ve run into (especially with there being some more-or-less duplicated subsystems, with the old being incrementally replaced by the new)

There was a Weeder config in the repo, but it used the older Dhall format instead of the newer TOML, and has a minimal-enough config that I’m not sure whether it was ever used, and certainly not recently.

Weeder is now also included in the Nix environment and run in CI.

## Implementation notes

This is mostly the first few commits from #5392, which was rightly closed.

## Loose ends

This PR doesn’t delete any code. All existing weeds are grandfathered in. #5392 attempted to do too much at once, and so nothing got in. This intends to stop the accumulation of weeds first, with removal of existing weeds to come later.

sellout committed:
Make IFD explicit in the flake

`allow-import-from-derivation` is true by default in Nix, but if a user
has turned it off (as I have recently), our flake will fail without this.

sellout committed:
Get Weeder working

Added Weeder to the default Nix shell.

Weeder no longer uses Dhall for its config, replaced it with an
equivalent TOML one.

sellout committed:
Exempt all existing weeds

This is temporary so that weeds can be tracked going forward without
having to fix all the existing ones first.

sellout committed:
Add Weeder to CI

sellout committed:
Document Weeder usage

sellout committed:
Bless modules whose APIs are independent of Unison

Also ones with debugging functions.

I don’t know why, but this caused some additional weeds to be discovered.

sellout committed:
Add remaining Weeder keys

To make it more explicit what isn’t being checked.

sellout committed:
Enable `unused-types` in Weeder

sellout committed:
Predefine some `root-instances` for Weeder

This doesn’t yet turn off `type-class-roots`, because whitelisting all
of those is a pain, but it pre-populates `root-instances` with instances
that we would want to keep, even if unused.

sellout committed:
Only run Weeder on x86_64-linux in CI

aryairani reviewed (approved):
I'm thinking it should go into check.sh too if it's going into CI?

sellout committed:
Add `check-weeds` script

This script also gets run as part of `scripts/check.sh`.

Since `check-weeds` has to do much of the same work as
`check-formatting`, I extracted the common code to a separate Bash file.

Also made it so the scripts don’t have to be run from the repo root
directory.

sellout committed:
Don’t run Weeder in CI if using cached binaries
